compaction/compaction_manager.hh

```diff
@@ -130,7 +130,8 @@ private:
 
         bool compaction_disabled() const;
     };
-    std::unordered_map<table*, compaction_state> _compaction_state;
+    using compaction_state_ptr = lw_shared_ptr<compaction_state>;
+    std::unordered_map<table*, compaction_state_ptr> _compaction_state;
 
```

Comment:
This seems like a big change just to stop new compactions.

Adding shared pointers makes things more difficult since lifetimes are less clear. We could do it, but should try for simpler solutions first.

Reply:
I'll see if I can simplify this series without using a shared pointer.  It might be possible with `std::unordered_map::detach` right away in `compaction_manager::remove` like done here, since it keeps the `compaction_state` object alive in the returned node handle.

Reply:
We already had compaction_state::gate and compaction_state::lock to synchronize with outstanding tasks.

If we also want to not depend on the lock FIFO order, we can have compaction tasks enter the compaction_state gate, in addition to functions running with compaction disabled that currently enter this gate.

And then compaction_manager::remove can just close this gate to wait on all tasks to complete and make sure they won't need compaction_state after that.

Reply:
remove() is the only one erasing compaction state for a table T. we only need to make sure that no task is left over, nor created while table is being removed. 

we probably need to add a state to compaction_state to prevent new tasks from being submitted while stopping tasks for T.

Reply:
`can_proceed` will check for the existence of _compaction_state[cf].
See in v2

Reply:
v2 implemented without shared ptr.
Instead tasks keep a reference to the compaction_state and enter its gate so that `remove()` will wait on all tasks and functions (running with compaction disabled) by closing the gate